File: leetcode006/main.go
package main

import "fmt"

func main() {
	nums := []int{-1, 1, 0, -3, 3}
	results := productExceptSelf(nums)
	fmt.Println(results)
}

func productExceptSelf(nums []int) []int {
	// create an array with items == 1 if length size of nums
	size := len(nums) // Array size
	results := make([]int, size)

	// prefix : stores product of all items before them
	prefix := 1
	// Loop through items in nums and get the prefixes
	for i := 0; i < len(nums); i++ {
		results[i] = prefix
		prefix *= nums[i]
	}
	// prefix : stores product of all items after them
	postfix := 1
	// Loop through items in nums and get the postfixes

	for i := len(nums) - 1; i >= 0; i-- {
		results[i] = postfix
		postfix *= nums[i]
	}
	return results
}

File: leetcode006/main.py
from typing import List


def productExceptSelf(nums: List[int]) -> List[int]:
    # create an array results with size of number of items in nums : all items will be 1 i.e [1,1,1, ..... len(nums)]
    result = [1] * len(nums)
    # prefix : product of individual items before them
    prefix = 1
    # Loop through items from left to right : prefixing
    for i in range(len(nums)):
        result[i] = prefix
        prefix *= nums[i]

    # postfix : product of individual items after them
    postfix = 1
    # Loop through items from right to left : postfixing
    for i in range(len(nums)-1,-1,-1):
        result[i] = postfix
        postfix *= nums[i]

    return result


if __name__ == "__main__":
    nums = [-1,1,0,-3,3]
    results = productExceptSelf(nums)
    print(results)

    


